polling/frequent/activities.py

```diff
@@ -16,10 +16,11 @@ class ComposeGreetingInput:
 @activity.defn
 async def compose_greeting(input: ComposeGreetingInput) -> str:
     test_service = TestService()
+    attempt = 1
     while True:
         try:
             try:
-                result = await test_service.get_service_result(input)
+                result = await test_service.get_service_result(input, attempt)
```

Comment:
I think that the way the sample previously did not pass the attempt number to the test service is desirable to maintain, since in real world examples the service in question won't want to know about our attempt counting.

Is there a way to fix the algorithm without explicitly passing the attempt number to the service?

Reply:
I understand and like the meta-point of wanting samples to look real-world, though this one feels shippable anyway.  I could use a global, but it wouldn't work out-of-process, and I'm not inspired to spend the time to change this.

Reply:
OK, I do think it's worth finding a way to make your test pass without making the sample misleading. Here's a global variable version: https://github.com/temporalio/samples-python/pull/152 (I'm starting to see why the code had that modulo-arithmetic return condition!)

Reply:
Yeah, exactly, this doesn't work predictably when it's run multiple times on the same worker.  I really don't think anybody's going to be misled by it as it is, and I would greatly prefer predictability and test isolation as a user. 
That said, I wouldn't block your approach if you feel strongly and that this is worth your time.
Can you either accept my PR and then add on top of it, or just leave it as I have it?

Reply:
GitHub was having problems with this PR after I brought it up-to-date with `main` and it's closed currently. Are you able to reopen it?

The code in this sample falls into two categories:

1. Code that we are providing to users, essentially saying "follow this closely; this is what your code should look like". Note that users may well copy and paste code in this category. **In particular**: `workflows.py`, `activities.py`  

2. Private implementation detail of the sample, which exists only to make the sample work. Users will not copy and paste this code because, in their use cases, its role will be played by something specific to their domain. **In particular**: `test_service.py`


For (2) the basic idea of the sample is to create a toy service that is stateful, only responding after a few attempts. But that statefulness needs to be private to the `test_service` implementation, since it's modeling real-world slowness-to-come-up. The one thing we don't want to do is give users the impression that they need to count their attempts in their activity code and send them to a service. The sample in its current form takes care not to do that, and we don't need to make the same worse in that respect.

There are a few options I think:

1. We could move the counter in  the `test_service` to the module level and use modulo arithmetic to address the issue that it might be used by multiple workflows.
2. We could move the line `test_service = TestService()` in `activities.py` to the top level and use modulo arithmetic to address the issue that it might be used by multiple workflows.
3. We could use a dict keyed by `activity.info().workflow_id` for the counter in `test_service`

All those would teach the lesson we're trying to teach correctly. (1) and (2) are suboptimal in that they'd result in some potentially confusing shared state if someone were to run multiple instances of the sample concurrently. So (3) is perhaps the best choice, seeing as it's hardly more complex than (1) and (2).

Reply:
I've implemented option (3) in https://github.com/temporalio/samples-python/pull/152, which builds upon this PR (i.e. using the test you added).